contracts/external/gmxV2/GmxV2IsolationModeTokenVaultV1.sol

```diff
@@ -0,0 +1,421 @@
+    // =========================== Constants ============================
+    // ==================================================================
+
+    bytes32 private constant _FILE = "GmxV2IsolationModeVaultV1";
+    bytes32 private constant _VIRTUAL_BALANCE_SLOT = bytes32(uint256(keccak256("eip1967.proxy.virtualBalance")) - 1);
+    bytes32 private constant _VAULT_FROZEN_SLOT = bytes32(uint256(keccak256("eip1967.proxy.vaultFrozen")) - 1);
+    bytes32 private constant _SOURCE_IS_WRAPPER_SLOT = bytes32(uint256(keccak256("eip1967.proxy.sourceIsWrapper")) - 1);
+    bytes32 private constant _SHOULD_SKIP_TRANSFER_SLOT = bytes32(uint256(keccak256("eip1967.proxy.shouldSkipTransfer")) - 1); // solhint-disable max-line-length 
+
+    // ===================================================
+    // ==================== Modifiers ====================
+    // ===================================================
+
+    modifier requireNotFrozen() {
+        Require.that(
+            !isVaultFrozen(),
+            _FILE,
+            "Vault is frozen"
+        );
+        _;
+    }
+    
+    // ==================================================================
+    // ======================== Public Functions ========================
+    // ==================================================================
+
+    function initiateWrapping(
+        uint256 _tradeAccountNumber,
+        uint256[] calldata _marketIdsPath,
+        uint256 _inputAmountWei,
+        uint256 _minOutputAmountWei,
+        IGenericTraderProxyV1.TraderParam[] memory _tradersPath,
+        IDolomiteMargin.AccountInfo[] memory _makerAccounts,
+        IGenericTraderProxyV1.UserConfig memory _userConfig
+    ) external payable nonReentrant onlyVaultOwner(msg.sender) requireNotFrozen() {
+        if (msg.value > 0) {
+            address payable wrapper = payable(address(registry().gmxV2WrapperTrader()));
+            (bool success, ) = wrapper.call{value: msg.value}("");
+            Require.that(
+                success,
+                _FILE,
+                "Unable to send funds to wrapper"
+            );
+        }
+
+        _swapExactInputForOutput(
+            _tradeAccountNumber,
+            _marketIdsPath,
+            _inputAmountWei,
+            _minOutputAmountWei,
+            _tradersPath,
+            _makerAccounts,
+            _userConfig
+        );
+        // @audit Will this allow reentrancy in _swapExactInputForOutput. May have to requireNotFrozen on external functions instead of internal
+        _setVaultFrozen(true);
+    }
+
+    function initiateUnwrapping() external payable nonReentrant onlyVaultOwner(msg.sender) requireNotFrozen() {
+        // @todo
+    }
+
+    // @audit Need to check this can't be used to unfreeze the vault with a dummy deposit. I don't think it can
+    function cancelDeposit(bytes32 _key) external onlyVaultOwner(msg.sender) {
+        registry().gmxV2WrapperTrader().cancelDeposit(_key);
```

Comment:
This assumes calling `cancelDeposit` reverts if the key is invalid. Make sure you comment/document that here.